source/audio.py

```diff
@@ -0,0 +1,210 @@
+
+VolumeTupleT = Tuple[float, float]
+
+
+@unique
+class SoundSplitState(DisplayStringIntEnum):
+	OFF = 0
+	NVDA_LEFT = 1
+	NVDA_RIGHT = 2
+
+	@property
+	def _displayStringLabels(self) -> Dict[IntEnum, str]:
+		return {
+			# Translators: Sound split state
+			SoundSplitState.OFF: _("Disabled sound split"),
+			# Translators: Sound split state
+			SoundSplitState.NVDA_LEFT: _("NVDA on the left and applications on the right"),
+			# Translators: Sound split state
+			SoundSplitState.NVDA_RIGHT: _("NVDA on the right and applications on the left"),
+		}
+	
+	def getAppVolume(self) -> VolumeTupleT:
+		return {
+			SoundSplitState.OFF: (1.0, 1.0),
+			SoundSplitState.NVDA_LEFT: (0.0, 1.0),
+			SoundSplitState.NVDA_RIGHT: (1.0, 0.0),
+		}[self]
+	
+	def getNVDAVolume(self) -> VolumeTupleT:
+		return {
+			SoundSplitState.OFF: (1.0, 1.0),
+			SoundSplitState.NVDA_LEFT: (1.0, 0.0),
+			SoundSplitState.NVDA_RIGHT: (0.0, 1.0),
+		}[self]
+
+
+@unique
+class SoundSplitToggleMode(DisplayStringIntEnum):
+	OFF_LEFT_RIGHT = 0
+	OFF_AND_NVDA_LEFT = 1
+	OFF_AND_NVDA_RIGHT = 2
+
+	@property
+	def _displayStringLabels(self) -> Dict[IntEnum, str]:
+		return {
+			# Translators: Sound split toggle mode
+			SoundSplitToggleMode.OFF_AND_NVDA_LEFT: _("Cycles through off and NVDA on the left"),
+			# Translators: Sound split toggle mode
+			SoundSplitToggleMode.OFF_AND_NVDA_RIGHT: _("Cycles through off and NVDA on the right"),
+			# Translators: Sound split toggle mode
+			SoundSplitToggleMode.OFF_LEFT_RIGHT: _("Cycles through off, NVDA on the left and NVDA on the right"),
+		}
+	
+	def getPossibleStates(self) -> List[SoundSplitState]:
+		result = [SoundSplitState.OFF]
+		if 'LEFT' in self.name:
+			result.append(SoundSplitState.NVDA_LEFT)
+		if 'RIGHT' in self.name:
+			result.append(SoundSplitState.NVDA_RIGHT)
+		return result
+	
+	def getClosestState(self, state: SoundSplitState) -> SoundSplitState:
+		states = self.getPossibleStates()
+		if state in states:
+			return state
+		return states[-1]
+
+
+sessionManager: audiopolicy.IAudioSessionManager2 = None
+activeCallback: Optional[comtypes.COMObject] = None
+
+
+def initialize() -> None:
+	global sessionManager
+	sessionManager = getSessionManager()
+	if sessionManager is None:
+		log.error("Could not initialize audio session manager! ")
+		return
+	if nvwave.usingWasapiWavePlayer():
+		state = SoundSplitState(config.conf['audio']['soundSplitState'])
+		global activeCallback
+		activeCallback = setSoundSplitState(state)
+
+
+@atexit.register
+def terminate():
+	if nvwave.usingWasapiWavePlayer():
+		setSoundSplitState(SoundSplitState.OFF)
+		if activeCallback is not None:
+			unregisterCallback(activeCallback)
+
+
+def getDefaultAudioDevice(kind: EDataFlow = EDataFlow.eRender) -> Optional[mmdeviceapi.IMMDevice]:
+	deviceEnumerator = comtypes.CoCreateInstance(
+		CLSID_MMDeviceEnumerator,
+		mmdeviceapi.IMMDeviceEnumerator,
+		comtypes.CLSCTX_INPROC_SERVER,
+	)
+	device = deviceEnumerator.GetDefaultAudioEndpoint(

```

Comment:
This currently fails on machines (such as the CI node on AppVeyor) where there are no audio devices, with the following error:
```
  File "audio.pyc", line 98, in initialize
  File "audio.pyc", line 130, in getSessionManager
  File "audio.pyc", line 122, in getDefaultAudioDevice
_ctypes.COMError: (-2147023728, 'Element not found.', (None, None, None, 0, None))
```

Reply:
I added try-catch clause downstream where I log this error.